A user can create a new allocated booking for a person

bryangaledxw opened the pull request:
# Changes in this PR

- A user can book a specific bedspace within a property

## Screenshots of UI changes

![localhost_3000_properties_52b62e02-8f72-4baf-b3b2-dd09d3ad395d_bedspaces_f1442ee5-baaf-450a-bcb0-057895a16d03_bookings_new](https://user-images.githubusercontent.com/94137563/203040011-6bab69cf-65bf-4cea-bd4a-b49431e89218.png)

bryangaledxw committed:
Add date and input Nunjucks components

These are borrowed from Approved Premises, and will make it easier to use GDS components in a way that follows our project's conventions

bryangaledxw committed:
Fix incorrect titles and unneeded imports in our bedspaces pages

bryangaledxw committed:
Correct the return code of our mocked API in our PremisesClient.create test

bryangaledxw committed:
Remove unneeded fields from NewBooking factory

Previously the factory was adding these fields to replicate data that would come from the browser, before being marshalled into a NewBooking. However, problems arose as the month was off-by-one (Javascript uses zero-indexed months in places), and we don't want to rely on this additional data in our testing

bryangaledxw committed:
Add getRoom to BedspaceService

We will want to display the bedspace name, which lives on the Room object, on the page for creating a booking

bryangaledxw committed:
Add createForBedspace to BookingService

We make it the BookingService's responsibility to be aware that to create a booking for a bedspace, we need to extract the bed ID from the bedspace's room

bryangaledxw committed:
Route our new booking page

We can now reach our new booking page by directly navigating to its URL. This exposes the page for integration testing, and gives us a destination for the page's form, and a redirect path when there are errors

bryangaledxw committed:
Link to our new booking page from the show bedspace page

This allows us to expand our integration testing, and exposes the page for E2E testing

bryangaledxw committed:
Add shared partial _editable.njk for our new booking page

We will want the same fields for both creating and editing a booking, so we create this shared partial, and associated abstract BookingEditable class that will help when we integration test these pages.

We also add the error messages we will want when this partial is used

bryangaledxw committed:
Add BookingsController and associated Nunjucks template

We also add a BookingNew helper class for integration testing this new page

Note we use the insertGenericError helper function we created earlier to handle 409 Conflict errors, that occur when the new booking conflicts with an existing booking

bryangaledxw committed:
Generate latest shared types

We also update our Booking and NewBooking factories to match these new types

bryangaledxw committed:
Extract bed factory

We will sometimes want to construct a Bed in our testing, to ensure that it has a particular bed ID, so we extract a bed factory to allow us to do this

bryangaledxw committed:
Remove unused ReferenceDataClient dependency from BookingService

bryangaledxw committed:
Add a premisesId constant to our BookingService tests

As with similar tests, we declare a premisesId constant near the start of the test file

bryangaledxw committed:
Tidy naming of our booking stubbing tasks for integration tests

We tidy the naming here to be closer to similar tasks for premises and rooms

bryangaledxw committed:
Tidy CSS and extract reusable components

We follow the convention of keeping CSS for reusable components in a components folder

bryangaledxw committed:
Add insertGenericError helper function

We will use this in cases where we want to handler API errors that don't follow the format of giving an "errorType" for a "propertyName". This will allow us to handle these errors using the same conventions as other errors. We will use this when the API returns a 409 Conflict error, which will occur if we try to create conflicting bookings

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [A user can create a new 'allocated' booking for a person](https://trello.com/c/KCMRQZEm/532-a-user-can-create-a-new-allocated-booking-for-a-person)

pezholio commented on server/utils/dateUtils.ts:
```diff
@@ -30,7 +30,7 @@ export class DateFormats {
       return format(date, 'cccc d MMMM y')
     } else {
-      return format(date, 'dd/LL/y')
+      return format(date, 'd MMM yy')
```
This is a good shout - I think it's more in line with GDS standards, but should it be in its own commit?

bryangaledxw replied:
Done

pezholio commented on server/utils/validation.ts:
```diff
@@ -72,6 +72,20 @@ export const errorMessage = (field: string, text: string): ErrorMessage => {
+  })
+
+  data['invalid-params'] = invalidParams
+  // eslint-disable-next-line dot-notation
```
Is there a reason we can't use dot notation here?

bryangaledxw replied:
Then it fails on the typecheck step, as `data` isn't a known part of the Error type

pezholio commented on server/services/bookingService.ts:
```diff
@@ -21,6 +28,18 @@ export default class BookingService {
+
+    const confirmedBooking = await bookingClient.create(premisesId, {
+      serviceName: 'temporary-accommodation',
+      bedId: room.beds[0].id,
```
What happens if a room doesn't have any beds?

bryangaledxw replied:
I feel ok with this just erroring. All TA rooms should have a bed, and if not, something has gone wrong somewhere. In the case that a room doesn't have a bed, this isn't something we can elegantly handle I feel ok with that erroring here.

pezholio replied:
👍

pezholio commented on server/views/temporary-accommodation/bookings/_editable.njk:
```diff
@@ -0,0 +1,54 @@
+{{ taInput(
+    {
+      label: {
+        text: "What is the person's CRN",
```
I dunno if it might be better from a user's POV if we ask for the person's CRN first, and then confirm it on that page? There might be a risk that this could be booked for the wrong person if the user can't confirm the person's details. Ignore me if this is coming in the future though!

pezholio replied:
Also, there might be unforseen circumstances if the person is a limited access offender, though I might be getting too deep in the weeds!

bryangaledxw replied:
Yep, we're planning to pull that step out, this is intentionally getting a quick "let us create a booking" in

pezholio replied:
Cool - I thought that might be the case 👍

pezholio commented on integration_tests/tests/temporary-accommodation/manage/booking.cy.ts:
```diff
@@ -0,0 +1,108 @@
+    const page = BookingNewPage.visit(premises.id, room.id)
+
+    // And I miss required fields
+    cy.task('stubBookingCreateErrors', {
```
Could we stub a 409 here as well? Just to confirm everything wires up OK

bryangaledxw replied:
Sorted

pezholio reviewed (approved):
Looks generally good. Just a couple of things to think about, though you may have already considered them!

pezholio commented on server/utils/validation.test.ts:
```diff
@@ -180,3 +185,53 @@ describe('fetchErrorsAndUserInput', () => {
   })
 })
+
+describe('insertPropertyError', () => {
```
Maybe `insertGenericError` would be better, as we already have the concept of `Properties`

bryangaledxw replied:
Was torn between `insertPropertyError` and `insertParameterError`, as we use both terms - happy to sidestep the question entirely!

bryangaledxw committed:
Add convertIsoToDateAndTimeInputs helper function

We will use this in our booking tests to go from a NewBooking object, to the input from the browser that would have created this object

bryangaledxw committed:
Change our "short" date format to match designs
